### PR TITLE
Fix regex parsing logic handling of nested quantifiers

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Fixes the libcudf regex parsing logic when handling nested fixed quantifiers. The logic handles fixed quantifiers by simple repeating the previous instruction. If the previous item is a group (capture or non-capture) that group may also contain an internal fixed quantifier as well.
Found while working on #16730

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Review by bdice — APPROVED** [review]

File: cpp/src/strings/regex/regcomp.cpp [hunk 1]
```diff
@@ -731,26 +731,39 @@ class regex_parser {
...
-        // minimum repeats (n)
-        for (int j = 1; j < n; j++) {
-          out.insert(out.end(), begin, end);
+        if (n == 0 && m == 0) { out.erase(begin, end); }
+
+        std::vector<regex_parser::Item> repeat_copy(begin, end);
+        // special handling for quantified capture groups
+        if ((n > 1) && (*begin).type == LBRA) {
+          (*begin).type = LBRA_NC;  // change first one to non-capture
+          // add intermediate groups as non-capture
```
Comment (vyasr) [inline cpp/src/strings/regex/regcomp.cpp:748]: Why do we make these non-capturing? Do we handle capturing in some other way that isn't local to this code so I'm not seeing it, or is there some reason that we're guaranteed that these will never be capturing groups (or is this just a case that we explicitly don't support)?

**Review by vyasr — COMMENTED** [review]
If I understand correctly, all of the in->out substitutions exist because in the nested case out could be filled with additional instructions (lines 746-759), so the sizes now depend on that. Is that right? I think I've grokked this PR but I want to confirm my understanding before I approve.

Out of scope for this PR, but should regcomp.h be renamed to regcomp.hpp? It's not a C header, there are classes in there.

**Comment by davidwendt** [comment]
That sounds right. Previously it was just copying the input and so did not account for previous expanded (nested) instructions.
I can look into renaming the header. The original implementation came from C source I believe.

**Review by vyasr — APPROVED** [review]

File: cpp/src/strings/regex/regcomp.cpp (hunk 1 above)
Comment (davidwendt) [reply cpp/src/strings/regex/regcomp.cpp:748]: The behavior is an attempt to honor the regex spec defined in the first paragraph here
https://www.regular-expressions.info/captureall.html
Specifically this phrase: _repeated capturing group will capture only the last iteration_
So all the intermediate groups are converted to non-capture leaving only the last iteration as capturing.
Example from #16730: https://regex101.com/r/QdDTVZ/1

**Pull request merged**